File: src/main/java/waterpower/common/block/tileentity/TileEntityInventory.java
package waterpower.common.block.tileentity;

import java.util.ArrayList;
import java.util.List;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.ISidedInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.TextComponentString;
import waterpower.common.block.inventory.InventorySlot;

public abstract class TileEntityInventory extends TileEntityBase implements ISidedInventory {
    private final List<InventorySlot> invSlots = new ArrayList<InventorySlot>();

    @Override
    public void readFromNBT(NBTTagCompound nbtTagCompound) {
        super.readFromNBT(nbtTagCompound);

        NBTTagCompound invSlotsTag = nbtTagCompound.getCompoundTag("InvSlots");

        for (InventorySlot invSlot : this.invSlots)
            invSlot.readFromNBT(invSlotsTag.getCompoundTag(invSlot.name));
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound nbtTagCompound) {
        super.writeToNBT(nbtTagCompound);

        NBTTagCompound invSlotsTag = new NBTTagCompound();

        for (InventorySlot invSlot : this.invSlots) {
            NBTTagCompound invSlotTag = new NBTTagCompound();

            invSlot.writeToNBT(invSlotTag);

            invSlotsTag.setTag(invSlot.name, invSlotTag);
        }

        nbtTagCompound.setTag("InvSlots", invSlotsTag);
        
        return nbtTagCompound;
    }

    public List<InventorySlot> getInventorySlots() {
        return invSlots;
    }

    @Override
    public int getField(int id) {
        return 0;
    }

    @Override
    public void setField(int id, int value) {
    }

    @Override
    public int getFieldCount() {
        return 0;
    }

    @Override
    public boolean hasCustomName() {
        return true;
    }

	@Override
	public ITextComponent getDisplayName() {
		return new TextComponentString(getName());
	}

    @Override
    public int getSizeInventory() {
        int ret = 0;
        for (InventorySlot invSlot : this.invSlots)
            ret += invSlot.size();
        return ret;
    }

    @Override
    public ItemStack getStackInSlot(int index) {
        for (InventorySlot invSlot : this.invSlots) {
            if (index < invSlot.size())
                return invSlot.get(index);
            index -= invSlot.size();
        }

        return null;
    }

    @Override
    public ItemStack decrStackSize(int index, int amount) {
        ItemStack itemStack = getStackInSlot(index);
        if (itemStack == null)
            return null;

        if (amount >= itemStack.stackSize) {
            setInventorySlotContents(index, null);

            return itemStack;
        }

        itemStack.stackSize -= amount;

        ItemStack ret = itemStack.copy();
        ret.stackSize = amount;

        return ret;
    }

    @Override
    public ItemStack removeStackFromSlot(int index) {
        ItemStack ret = getStackInSlot(index);

        if (ret != null)
            setInventorySlotContents(index, null);

        return ret;
    }

    @Override
    public void setInventorySlotContents(int index, ItemStack itemStack) {
        for (InventorySlot invSlot : this.invSlots) {
            if (index < invSlot.size()) {
                invSlot.put(index, itemStack);
                break;
            }
            index -= invSlot.size();
        }
    }

    public boolean isInvNameLocalized() {
        return false;
    }

    @Override
    public int getInventoryStackLimit() {
        return 64;
    }

    @Override
    public boolean isUseableByPlayer(EntityPlayer entityPlayer) {
        return entityPlayer.getDistanceSq(pos.add(0.5, 0.5, 0.5)) <= 64.0D;
    }

    @Override
    public void openInventory(EntityPlayer player) {
    }

    @Override
    public void closeInventory(EntityPlayer player) {
    }

    @Override
    public boolean isItemValidForSlot(int index, ItemStack itemStack) {
        InventorySlot invSlot = getInvSlot(index);

        return (invSlot != null) && (invSlot.canInput()) && (invSlot.accepts(itemStack));
    }

    @Override
    public int[] getSlotsForFace(EnumFacing side) {
        int[] ret = new int[getSizeInventory()];

        for (int i = 0; i < ret.length; i++)
            ret[i] = i;

        return ret;
    }

    @Override
    public boolean canInsertItem(int index, ItemStack itemStack, EnumFacing side) {
        InventorySlot targetSlot = getInvSlot(index);
        if (targetSlot == null)
            return false;

        if ((!targetSlot.canInput()) || (!targetSlot.accepts(itemStack)))
            return false;
        if ((targetSlot.preferredSide != InventorySlot.InvSide.ANY) && (targetSlot.preferredSide.matches(side)))
            return true;

        for (InventorySlot invSlot : this.invSlots) {
            if ((invSlot != targetSlot) && (invSlot.preferredSide != InventorySlot.InvSide.ANY) && (invSlot.preferredSide.matches(side))
                    && (invSlot.canInput()) && (invSlot.accepts(itemStack))) {
                return false;
            }
        }

        return true;
    }

    @Override
    public boolean canExtractItem(int index, ItemStack itemStack, EnumFacing side) {
        InventorySlot targetSlot = getInvSlot(index);
        if (targetSlot == null)
            return false;

        if (!targetSlot.canOutput())
            return false;
        if ((targetSlot.preferredSide != InventorySlot.InvSide.ANY) && (targetSlot.preferredSide.matches(side)))
            return true;

        for (InventorySlot invSlot : this.invSlots) {
            if ((invSlot != targetSlot) && (invSlot.preferredSide != InventorySlot.InvSide.ANY) && (invSlot.preferredSide.matches(side))
                    && (invSlot.canOutput())) {
                return false;
            }
        }

        return true;
    }

    public void addInvSlot(InventorySlot invSlot) {
        this.invSlots.add(invSlot);
    }

    private InventorySlot getInvSlot(int index) {
        for (InventorySlot invSlot : this.invSlots) {
            if (index < invSlot.size()) {
                return invSlot;
            }
            index -= invSlot.size();
        }

        return null;
    }

    @Override
    public void clear() {
        for (InventorySlot invSlot : this.invSlots)
            invSlot.clear();
    }

}

File: src/main/kotlin/waterpower/common/block/tile/TileEntityInventory.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.tile

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.ISidedInventory
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraft.util.text.TextComponentString
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.items.CapabilityItemHandler
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.wrapper.InvWrapper
import net.minecraftforge.items.wrapper.SidedInvWrapper
import waterpower.common.block.inventory.InventorySlot
import waterpower.util.*
import java.util.*

abstract class TileEntityInventory : TileEntityBase(), ISidedInventory {
    private val itemHandler = arrayOfNulls<IItemHandler>(EnumFacing.VALUES.size + 1)
    val invSlots = ArrayList<InventorySlot>()

    override fun readFromNBT(tag: NBTTagCompound) {
        super.readFromNBT(tag)

        val inventory = tag.getCompoundTag("inventory")

        for (invSlot in this.invSlots)
            invSlot.readFromNBT(inventory.getCompoundTag(invSlot.name))
    }

    override fun writeToNBT(tag: NBTTagCompound): NBTTagCompound {
        super.writeToNBT(tag)

        val invSlotsTag = NBTTagCompound()
        for (invSlot in invSlots) {
            val invSlotTag = NBTTagCompound()
            invSlot.writeToNBT(invSlotTag)
            invSlotsTag.setTag(invSlot.name, invSlotTag)
        }

        tag.setTag("inventory", invSlotsTag)

        return tag
    }

    override fun hasCustomName() = true

    override fun isEmpty(): Boolean {
        for (invSlot in this.invSlots) {
            if (!invSlot.isEmpty()) {
                return false
            }
        }
        return true
    }

    override fun getSizeInventory(): Int {
        var ret = 0
        for (invSlot in this.invSlots)
            ret += invSlot.size()
        return ret
    }

    override fun canInsertItem(index: Int, itemStack: ItemStack, side: EnumFacing): Boolean {
        if (isStackEmpty(itemStack)) return false
        val targetSlot = getInvSlot(index) ?: return false

        if (!targetSlot.canInput() || !targetSlot.accepts(itemStack))
            return false
        if (targetSlot.preferredSide !== InventorySlot.InvSide.ANY && targetSlot.preferredSide.matches(side))
            return true

        for (invSlot in this.invSlots) {
            if (invSlot != targetSlot && invSlot.preferredSide !== InventorySlot.InvSide.ANY && invSlot.preferredSide.matches(side)
                    && invSlot.canInput() && invSlot.accepts(itemStack)) {
                return false
            }
        }

        return true
    }

    override fun canExtractItem(index: Int, itemStack: ItemStack, side: EnumFacing): Boolean {
        val targetSlot = getInvSlot(index) ?: return false

        if (!targetSlot.canOutput())
            return false
        if (targetSlot.preferredSide !== InventorySlot.InvSide.ANY && targetSlot.preferredSide.matches(side))
            return true

        for (invSlot in this.invSlots) {
            if (invSlot != targetSlot && invSlot.preferredSide !== InventorySlot.InvSide.ANY && invSlot.preferredSide.matches(side)
                    && invSlot.canOutput()) {
                return false
            }
        }

        return true
    }

    override fun removeStackFromSlot(index: Int): ItemStack {
        val ret = getStackInSlot(index)
        if (!isStackEmpty(ret))
            setInventorySlotContents(index, emptyStack)
        return ret
    }

    override fun getStackInSlot(index: Int): ItemStack {
        var i = index
        for (invSlot in this.invSlots) {
            if (i < invSlot.size())
                return invSlot[i]
            i -= invSlot.size()
        }

        return emptyStack
    }

    override fun getDrops(): List<ItemStack> {
        val list = ArrayList(super.getDrops())
        for (slot in invSlots)
            for (stack in slot)
                if (!isStackEmpty(stack))
                    list += stack
        return list
    }

    override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        if (capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
            if (facing == null) {
                if (itemHandler[itemHandler.size - 1] == null) {
                    itemHandler[itemHandler.size - 1] = InvWrapper(this)
                }
                return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast<T>(this.itemHandler[this.itemHandler.size - 1])
            }
            if (itemHandler[facing.ordinal] == null) {
                itemHandler[facing.ordinal] = SidedInvWrapper(this, facing)
            }
            return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.cast<T>(this.itemHandler[facing.ordinal])
        }
        return super.getCapability(capability, facing)
    }

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?) =
            capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing)

    override fun getSlotsForFace(side: EnumFacing?)
            = IntArray(sizeInventory, { it })

    override fun isItemValidForSlot(index: Int, stack: ItemStack): Boolean {
        val invSlot = getInvSlot(index)
        return invSlot != null && invSlot.canInput() && invSlot.accepts(stack)
    }

    override fun getInventoryStackLimit(): Int {
        var max = 0
        for (slot in this.invSlots) {
            max = maxOf(max, slot.stackSizeLimit)
        }
        return max
    }

    override fun openInventory(player: EntityPlayer?) {}
    override fun closeInventory(player: EntityPlayer?) {}

    override fun isUsableByPlayer(player: EntityPlayer?) =
            !isInvalid && player!!.getDistanceSq(pos) <= 64

    override fun markDirty() {
        super.markDirty()

        for (invSlot in invSlots)
            invSlot.markDirty()
    }

    override fun setInventorySlotContents(index: Int, stack: ItemStack) {
        var ret = index
        for (invSlot in invSlots)
            if (ret < invSlot.size()) {
                invSlot.put(ret, stack)
            } else
                ret -= invSlot.size()
    }

    override fun decrStackSize(index: Int, count: Int): ItemStack {
        var itemStack = getStackInSlot(index)
        if (isStackEmpty(itemStack)) return emptyStack

        if (count >= getCount(itemStack)) {
            setInventorySlotContents(index, emptyStack)
            return itemStack
        }

        itemStack = shrink(itemStack)
        return itemStack.copyWithNewCount(count)
    }

    fun addInvSlot(invSlot: InventorySlot) =
            this.invSlots.add(invSlot)

    private fun getInvSlot(index: Int): InventorySlot? {
        var i = index
        for (invSlot in this.invSlots) {
            if (i < invSlot.size()) {
                return invSlot
            }
            i -= invSlot.size()
        }

        return null
    }

    override fun clear() {
        for (invSlot in this.invSlots)
            invSlot.clear()
    }

    override fun getDisplayName() = TextComponentString(name)

    override fun getField(id: Int) = 0
    override fun setField(id: Int, value: Int) {}
    override fun getFieldCount() = 0
}